### ISSUE TITLE
move zipstream improvements to python

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7558]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#278|
|Type|task|
|Created|2003-09-24 21:32:52Z|

<details><summary>Searchable metadata</summary>

```
trac-id__278 278
type__task task
reporter__itamarst itamarst
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__worksforme worksforme
component__core core
keywords__ 
time__1064439172000000 1064439172000000
changetime__1467290614308530 1467290614308530
version__ 
owner__ 
cc__moonfallen cc__itamarst cc__kmike cc__exarkun cc__thijs cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Submit a patch with docs and tests to python's zipfile so at
least by 2.4 the functionality will be in stdlib.

</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moonfallen commented|
|-|-|

```
#!html
<pre>
I suggest that only the three classes be moved into standard Python and that
count* and unzip* be Twisted-specific.  

Honestly, I'm not sure I want to patch zipfile at all.  As someone (I think it
was z3p?) pointed out, zipfile.py is a mess that could stand a complete rewrite.
 Patching it seems like putting a new liver in a terminal cancer patient...
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
It's not *that* bad. And stdlib users needn't have to suffer.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @itamarst|
|-|-|

This didn't have tests last time we submitted it to the stdlib and it was rejected.  It has tests now, we should submit it again.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

And if this had been done, we wouldn't have had to fix it now to work with python 2.5.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Python bug tracker bug (which is currently in closed state for lack of tests):

http://sourceforge.net/tracker/index.php?func=detail&aid=424856&group_id=5470&atid=305470

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

That link is now http://bugs.python.org/issue424856

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

Maybe resubmit? I think the code is (slightly) better now.

**Comment by twisted-trac** [comment]
|<img alt="kmike's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|kmike commented|
|-|-|

Actually there is support for an incremental approach to unzipping files in zipfile since python 2.6.  ZipFile.open() method returns file-like object which supports 'read' method with buffer size in parameters.

Moreover, implementation is better than in zipstream.py. At least it works a lot faster because it don't have 'self.buffer = self.buffer[n:]' code which kills performance on big (even tens of Kb's) buffer sizes like zipstream version does.

I think is better to deprecate zipstream. Or at least copy implementation from new zipfile. Or to make zipstream a backported zipfile wrapper.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

So if the stdlib zipfile is now feature-equivalent to zipstream, then there's no reason for us to contribute anything upstream, and we should close this ticket.

I'm not really all that familiar with zipfile _or_ zipstream.  Is the former really feature-equivalent to the latter now?

**Comment by twisted-trac** [comment]
|<img alt="kmike's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|kmike commented|
|-|-|

zipfile does not have these:

```
unzip
unzipIter
countZipFileChunks
countFileChunks
countZipFileEntries
unzipIterChunky
```
But there is already [#3666](https://github.com/twisted/twisted/issues/3666) for these functions. 

ZipExtFile from zipfile doesn't provide 'tell' and 'isatty' methods that present in _FileEntry from zipstream but I think it is a very minor issue. I don't know the purpose of 'isatty' method. 'tell' method is used in zipstream to determine if all data is read but there are another ways to do it, one can just check if returned data is not empty. 

Besides that they are feature-equivalent.

The problem is ```ZipExtFile.read``` method is undocumented and I found that it is capable of uncompressing files without loading them into memory just occasionally while reading zipfile sources.

**Comment by twisted-trac** [comment]
|<img alt="kmike's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|kmike commented|
|-|-|

Actually ```ZipExtFile.read``` is mentioned in python docs but link incorrectly points to ```ZipFile.read``` method which has different purpose and signature.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

Sounds like it should be deprecated.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set status to `closed`|
|-|-|

Replying to [exarkun](#note_10):
> So if the stdlib zipfile is now feature-equivalent to zipstream, then there's no reason for us to contribute anything upstream, and we should close this ticket.
> 
> I'm not really all that familiar with zipfile _or_ zipstream.  Is the former really feature-equivalent to the latter now?

The main interesting function is `zipfile.ZIpFile.open`, [which is new in Python 2.6](http://docs.python.org/library/zipfile.html#zipfile.ZipFile.open).  We need something like this to implement `ZipPath.open`, which is really the only interesting thing in `zipstream` as far as I'm concerned.

We can't deprecate it yet because this feature is only in Python 2.6; we can drop it when we stop supporting 2.4 and 2.5, but that's a long ways off.

Closing as 'worksforme' because we didn't exactly "fix" this (AFAIK it's not our code that eventually got accepted) but the new feature works for me :).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|GitHub commented|
|-|-|

In changeset 79d61a975ff1ab0b18dc572a108f497168c63dfb
```
#!CommitTicketReference repository="" revision="79d61a975ff1ab0b18dc572a108f497168c63dfb"
Merge pull request #278:  Fix epydoc warnings in twisted/web

Author: rodrigc
Reviewer: adiroiban
Fixes: #8537
```
